code/datums/callback.dm

```diff
@@ -100,3 +100,60 @@
 	if (object == GLOBAL_PROC)
 		return call(delegate)(arglist(calling_arguments))
 	return call(object, delegate)(arglist(calling_arguments))
+
+
+/datum/callback_select
+	var/list/finished
+	var/pendingcount
+	var/total
+
+/datum/callback_select/New(count, savereturns)
+	total = count
+	if (savereturns)
+		finished = new(count)
+
+
+/datum/callback_select/proc/invoke_callback(index, datum/callback/callback, list/callback_args, savereturn = TRUE)
+	set waitfor = FALSE
+	if (!callback || !istype(callback))
+		//This check only exists because the alternative is callback_select would block forever if given invalid data
+		CRASH("invalid callback passed to invoke_callback")
+	if (!length(callback_args))
+		callback_args = list()
+	pendingcount++
+	debug_usr("calling callback")
+	var/rtn = callback.Invoke(arglist(callback_args))
+	debug_usr("callback returned")
+	pendingcount--
+	if (savereturn)
+		finished[index] = rtn
+
+
+
+
+//runs a list of callbacks asynchronously, returning once all of them return.
+//callbacks can be repeated.
+//callbacks-args is a optional list of argument lists, in the same order as the callbacks,
```

Comment:
select is a pretty standard programming syntax for a call that blocks on a set of objects and then returns when they're all ready/complete

although usually it's applied in the network programming world for socket io